### PR TITLE
Feature/smr api

### PR DESCRIPTION
**Pull request by kpatrashanu**

### CODE REVIEW

File: Services/Smr/Smr.API/Abstraction/Models/IInfoSmr.cs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        /// </value>
+        public string Subject { get; set; }
+
+        /// <summary>
+        /// Gets or sets the external identifier.
+        /// </summary>
+        /// <value>
+        /// The external identifier.
+        /// </value>
+        public string ExternalId { get; set; }
```
Comment (Anny777) [inline Services/Smr/Smr.API/Abstraction/Models/IInfoSmr.cs:121]: Rename on NotificationNumber

File: Services/Smr/Smr.API/Controllers/InfoSmrController.cs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+            {
+                return this.BadRequest($"Objects was not created.");
+            }
+
+            try
+            {
+                var createdObjectIds = await this.CreateAsync(infoSmrDtos).ConfigureAwait(false);
+                if (createdObjectIds.Any())
+                {
+                    return this.CreatedAtAction(string.Empty, createdObjectIds);
```
Comment (Anny777) [inline Services/Smr/Smr.API/Controllers/InfoSmrController.cs:63]: Get(Guid id) without string.Empty.

File: Services/Smr/Smr.API/Controllers/InfoSmrController.cs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            catch (UniqueIndexViolationException)
+            {
+                this.logger.LogWarning("UniqueIndexViolationException! Try to recreate object");
+                var createdObjectIds = await this.CreateAsync(infoSmrDtos).ConfigureAwait(false);
+                if (createdObjectIds.Any())
+                {
+                    return this.CreatedAtAction(string.Empty, createdObjectIds);
+                }
+
+                return this.Conflict("The object exist");
```
Comment (Anny777) [inline Services/Smr/Smr.API/Controllers/InfoSmrController.cs:75]: Not good idea. Need to discuss

File: Services/Smr/Smr.API/Migrations/20210408153736__InitialCreate.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    /// <summary>
+    /// Initial create.
+    /// </summary>
+    /// <seealso cref="Migration" />
+    public partial class InitialCreate : Migration
+    {
+#pragma warning disable
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.CreateTable(
```
Comment (Anny777) [inline Services/Smr/Smr.API/Migrations/20210408153736__InitialCreate.cs:19]: All properties must be not nullable.

File: Services/Smr/Smr.API/Services/InfoSmrRepository.cs [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+                {
+                    throw new UniqueIndexViolationException(e.Message, e);
+                }
+
+                throw new InvalidOperationException(e.Message, e);
+            }
+        }
+
+        /// <inheritdoc/>
+        public async Task<InfoSmr> GetAsync(Guid id)
```
Comment (Anny777) [inline Services/Smr/Smr.API/Services/InfoSmrRepository.cs:64]: Rename FindAsync

File: Services/Smr/Smr.API/Services/InfoSmrRepository.cs [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+            return await this.db.InfoSmr.FindAsync(id).ConfigureAwait(false);
+        }
+
+        /// <inheritdoc/>
+        public async Task<LoadResult> GetListAsync(DataSourceLoadOptions loadOptions)
+        {
+            return await DataSourceLoader.LoadAsync(this.db.InfoSmr, loadOptions).ConfigureAwait(false);
+        }
+
+        private static void Validate(IEnumerable<InfoSmrDto> infoSmrDto)
```
Comment (Anny777) [inline Services/Smr/Smr.API/Services/InfoSmrRepository.cs:75]: infoSmrDtos instead infoSmrDto

File: Services/Smr/Smr.API/Services/InfoSmrRepository.cs [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+                throw new InvalidOperationException(nameof(infoSmrDto));
+            }
+        }
+
+        private async Task<List<InfoSmr>> MapFromDto(IEnumerable<InfoSmrDto> infoSmrDtos)
+        {
+            var listInfoSmr = new List<InfoSmr>();
+            foreach (var infoSmrDto in infoSmrDtos)
+            {
+                var isExist = await this.FindAsync(infoSmrDto.ExternalId).ConfigureAwait(false);
```
Comment (Anny777) [inline Services/Smr/Smr.API/Services/InfoSmrRepository.cs:93]: Maybe subtract existing ones with one request, out of loop.

**Review by Anny777 — CHANGES_REQUESTED** [review]

File: Services/Smr/Smr.API/Abstraction/Models/IInfoSmr.cs [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        /// </summary>
+        public Guid Id { get; set; }
+
+        /// <summary>
+        /// Gets or sets the URL.
+        /// </summary>
+        /// <value>
+        /// The URL.
+        /// </value>
+        public string Url { get; set; }
```
[the remark this answers is not in the document]
Comment (DamienVStone) [reply Services/Smr/Smr.API/Abstraction/Models/IInfoSmr.cs:25]: Name should be Uri as well

File: Services/Smr/Smr.API/Abstraction/Models/IInfoSmr.cs [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+        /// </value>
+        public string Winner { get; set; }
+
+        /// <summary>
+        /// Gets or sets the start date.
+        /// </summary>
+        /// <value>
+        /// The start date.
+        /// </value>
+        public DateTime StartDate { get; set; }
```
Comment (Anny777) [inline Services/Smr/Smr.API/Abstraction/Models/IInfoSmr.cs:81]: Deleted unused property

File: Services/Smr/Smr.API/Abstraction/Models/IInfoSmr.cs [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+
+    /// <summary>
+    /// Smr info.
+    /// </summary>
+    public interface IInfoSmr
+    {
+        /// <summary>
+        /// Gets or sets the identifier.
+        /// </summary>
+        public Guid Id { get; set; }
```
Comment (Anny777) [inline Services/Smr/Smr.API/Abstraction/Models/IInfoSmr.cs:17]: Added DateTime CreatedDate property

**Review by Anny777 — CHANGES_REQUESTED** [review]

**Review by Anny777 — APPROVED** [review]

**Pull request merged**